### PR TITLE
Schema update for an array of strings

### PR DESCRIPTION
**Pull request by bpathak-ons**
### What

* Schema change for an array of strings

### How to review

* check if code make sense
* build, audit, unit works fine

### Who can review

* anyone not me

### CODE REVIEW

**Review by nshumoogum — DISMISSED** [review]

File: event/producer_test.go [hunk 1]
```diff
@@ -63,13 +62,12 @@ func TestProducer_SearchDataImport(t *testing.T) {
...
+			err := searchDataImportProducer.SearchDataImport(ctx, expectedSearchDataImportEvent)
 			So(err, ShouldBeNil)
 
 			var avroBytes []byte
 			select {
 			case avroBytes = <-pChannels.Output:
-				fmt.Printf("avroBytes: %v\n", avroBytes)
 				log.Event(ctx, "avro byte sent to producer output", log.INFO)
 			case <-time.After(time.Second * 5):
 				log.Event(ctx, "failing test due to timed out", log.INFO)
```
Comment (nshumoogum) [inline event/producer_test.go:73]: I know you didn't add this log line in this PR but would you mind improving it and basically combining it with t.FailNow in line below to `t.Fatalf("failing test due to timing out after %v seconds", timeoutAfter)` where `var testTimeout = 5 *time.Second` - can reuse variable in select statement like so: `case <-time.After(testTimeout):`

Comment (nshumoogum) [reply event/producer_test.go:73]: General rule when writing tests use t.log to logout messages and remember you can always combine a log with forcing the test fail.

File: go.mod [hunk 2]
```diff
@@ -8,12 +8,10 @@ require (
 	github.com/ONSdigital/dp-api-clients-go/v2 v2.2.0
 	github.com/ONSdigital/dp-component-test v0.3.0
 	github.com/ONSdigital/dp-healthcheck v1.1.0
-	github.com/ONSdigital/dp-kafka/v2 v2.2.0
+	github.com/ONSdigital/dp-kafka/v2 v2.4.2
```
Comment (nshumoogum) [inline go.mod:11]: It is worth updating to use latest go version 1.17 - I would advise updating your go version locally

If you make this change now, it is also worth updating ci files to use the latest go version too

File: handler/handler_test.go [hunk 3]
```diff
@@ -60,12 +60,13 @@ var (
...
 
 func TestContentPublishedHandler_Handle(t *testing.T) {
-	expectedSearchDataImport := marshalSearchDataImport(t, searchDataImportEvent)
 
 	kafkaProducerMock := &kafkatest.IProducerMock{
 		ChannelsFunc: getChannelFunc,
 	}
 
+	//for mock marshaller
+	expectedSearchDataImport := marshalSearchDataImport(t, expectedSearchDataImportEvent)
```
Comment (nshumoogum) [inline handler/handler_test.go:69]: It's worth having a test for when the data returned from Zebedee contains optional fields too:

1) All fields are returned from Zebedee
2) Only mandatory fields are returned from Zebedee

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: handler/handler_test.go (hunk 3 above)
Comment (bpathak-ons) [reply handler/handler_test.go:69]: tests updated for both scenarios

File: event/producer_test.go (hunk 1 above)
Comment (bpathak-ons) [reply event/producer_test.go:73]: Thanks. done the changes

File: handler/handler_test.go (hunk 3 above)
Comment (bpathak-ons) [reply handler/handler_test.go:69]: tests updated to include both scenarios

File: go.mod (hunk 2 above)
Comment (bpathak-ons) [reply go.mod:11]: go version 1.17.1 updated for mod and ci files

File: handler/handler_test.go [hunk 4]
```diff
@@ -59,13 +47,27 @@ var (
 	}
 )
 
-func TestContentPublishedHandler_Handle(t *testing.T) {
-	expectedSearchDataImport := marshalSearchDataImport(t, searchDataImportEvent)
+func TestHandlerForZebedeeReturingMandatoryFields(t *testing.T) {
```
Comment (nshumoogum) [inline handler/handler_test.go:50]: Typo - missing "n" in "Returing" => "Returning":
```suggestion
func TestHandlerForZebedeeReturningMandatoryFields(t *testing.T) {
```

File: handler/handler_test.go [hunk 5]
```diff
@@ -126,6 +134,83 @@ func TestContentPublishedHandler_Handle(t *testing.T) {
 	})
 }
 
+func TestHandlerForZebedeeReturingAllFields(t *testing.T) {
+
+	expectedFullSearchDataImportEvent := models.SearchDataImport{
+		DataType:        "testDataType",
+		JobID:           "JobId",
```
Comment (nshumoogum) [inline handler/handler_test.go:141]: For now this will be empty

File: handler/handler_test.go [hunk 6]
```diff
@@ -126,6 +134,83 @@ func TestContentPublishedHandler_Handle(t *testing.T) {
...
+			Convey("And then get content published from zebedee", func() {
+				So(zebedeeMock.GetPublishedDataCalls(), ShouldNotBeEmpty)
+				So(zebedeeMock.GetPublishedDataCalls(), ShouldHaveLength, 1)
+				So(zebedeeMock.GetPublishedDataCalls()[0].UriString, ShouldEqual, testEvent.URI)
+			})
+			Convey("And then the expected bytes are sent to producer.output", func() {
+				var actual models.SearchDataImport
+				err = schema.SearchDataImportEvent.Unmarshal(avroBytes, &actual)
+				So(err, ShouldBeNil)
+				So(expectedFullSearchDataImportEvent.DataType, ShouldResemble, actual.DataType)
```
Comment (nshumoogum) [inline handler/handler_test.go:205]: Again I would suggest replacing `ShouldResemble` with `ShouldEqual`, it is more precise.

File: handler/handler_test.go [hunk 7]
```diff
@@ -126,6 +134,83 @@ func TestContentPublishedHandler_Handle(t *testing.T) {
...
+				So(expectedFullSearchDataImportEvent.JobID, ShouldResemble, actual.JobID)
+				So(expectedFullSearchDataImportEvent.Keywords, ShouldResemble, actual.Keywords)
```
Comment (nshumoogum) [inline handler/handler_test.go:207]: Missing some assertions to cover optional fields

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: event/producer_test.go [hunk 8]
```diff
@@ -63,24 +62,24 @@ func TestProducer_SearchDataImport(t *testing.T) {
...
-			err := searchDataImportProducer.SearchDataImport(ctx, searchDataImportTestEvent)
+			err := searchDataImportProducer.SearchDataImport(ctx, expectedSearchDataImportEvent)
 			So(err, ShouldBeNil)
 
 			var avroBytes []byte
+			var testTimeout = time.Second * 5
 			select {
 			case avroBytes = <-pChannels.Output:
-				fmt.Printf("avroBytes: %v\n", avroBytes)
 				log.Event(ctx, "avro byte sent to producer output", log.INFO)
```
Comment (nshumoogum) [inline event/producer_test.go:72]: Sorry just noticed this one - replace log.Event with t.Log?

```suggestion
				t.log("avro byte sent to producer output")
```

File: handler/handler_test.go [hunk 9]
```diff
@@ -85,10 +87,12 @@ func TestContentPublishedHandler_Handle(t *testing.T) {
 			err := eventHandler.Handle(context.Background(), &config.Config{}, &testEvent)
 
 			var avroBytes []byte
+			var testTimeout = time.Second * 5
 			select {
 			case avroBytes = <-pChannels.Output:
-				fmt.Printf("avroBytes: %v\n", avroBytes)
-			case <-time.After(time.Second * 5):
+				log.Event(ctx, "avro byte sent to producer output", log.INFO)
```
Comment (nshumoogum) [inline handler/handler_test.go:93]: And again use `t.Log` instead of `log.Event`:
```suggestion
				t.log("avro byte sent to producer output")
```

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: handler/handler_test.go [hunk 10]
```diff
@@ -104,7 +108,11 @@ func TestContentPublishedHandler_Handle(t *testing.T) {
 				var actual models.SearchDataImport
 				err = schema.SearchDataImportEvent.Unmarshal(avroBytes, &actual)
 				So(err, ShouldBeNil)
-				So(searchDataImportEvent, ShouldResemble, actual)
+				So(expectedSearchDataImportEvent.DataType, ShouldResemble, actual.DataType)
+				So(expectedSearchDataImportEvent.JobID, ShouldResemble, actual.JobID)
+				So(expectedSearchDataImportEvent.Keywords, ShouldResemble, actual.Keywords)
+				So(expectedSearchDataImportEvent.Title, ShouldResemble, actual.Title)
```
[the remark this answers is not in the document]
Comment (bpathak-ons) [reply handler/handler_test.go:114]: updated with assertion for expected and actual as null string is still ''

**Review by nshumoogum — APPROVED** [review]

**Pull request merged**